### PR TITLE
Implement AccessTokenVerifier class

### PR DESCRIPTION
**Pull request by jtemporal**
### Changes

- Class added: `AccessTokenVerifier`

This class implements all the logic in our decorators for Python API quick starts. This aims to solve the issue of having to copy more than 150 lines of code, now instead people would install the python SDK and follow the example below in their endpoints/methods:

```python
from auth0.v3.authentication.token_verifier import (
    AccessTokenVerifier, SymmetricSignatureVerifier
)

from auth0.v3.exceptions import Auth0Error, TokenValidationError

sig_verifier = SymmetricSignatureVerifier('secret')

expected_issuer = 'https://tokens-test.auth0.com/'

api_audience = 'http://audience.test'

token_verifier = AccessTokenVerifier(
    signature_verifier=sig_verifier,
    issuer=expected_issuer,
    audience=api_audience
)

token = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJhdXRoMHxzZGs0NThma3MiLCJhdWQiOiJodHRwOi8vYXVkaWVuY2UudGVzdCIsImlzcyI6Imh0dHBzOi8vdG9rZW5zLXRlc3QuYXV0aDAuY29tLyIsImV4cCI6MTY4Nzc2NTM2MSwiaWF0IjoxNTg3NTkyNTYxLCJzY29wZSI6InNjb3BlOm9uZSBzY29wZTp0d28iLCJwZXJtaXNzaW9ucyI6WyJzY29wZTp0d28iXX0.AYakukIxuyfIuPxVAXGYmmLMHiBiahEBcSStjWr07BE'

# No error should be raised here
try:
    token_verifier.verify(token, permissions=['scope:two'], scopes='scope:one')
except TokenValidationError as error:
    print(error)
finally:
    print('Verification successful! Correct Permissions and scopes')

# The Auth0Error for wrong permissions should be raised
try:
    token_verifier.verify(token, permissions=['scope:three'])
except Auth0Error as error:
    print(error)

# The Auth0Error for wrong scopes should be raised
try:
    token_verifier.verify(token, scopes='scope:three')
except Auth0Error as error:
    print(error)

# Results
# Verification successful! Correct Permissions and scopes
# 403: You don't have access to this resource
# 403: Insufficient scope.
```

- Any alternative designs or approaches considered: I copied most of the code from the `TokenVerifier` class and I believe a refactor for avoiding code repetition would be something good to do, but I didn't want to do it in this PR as per the [Auth0 Contribution Guide directive on Pull Requests](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md#pull-requests).

### References

This is based on me copying code for the quickstarts for every sample I did so far, so this creates a class that can hold all the logic for validating an Auth0 Access Token

### Testing

Please describe how this can be tested by reviewers. Be specific about anything not tested and reasons why. If this library has unit and/or integration testing, tests should be added for new functionality and existing tests should complete without errors. 

- [x] This change adds unit test coverage
- [ ] This change adds integration test coverage (I'm not sure how to run those so I left it unchecked)
- [x] This change has been tested on the latest version of the platform/language

### Checklist

- [x] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)
- [x] I have read the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md)
- [x] All existing and new tests complete without errors

Screen Shot 2021-08-16 at 15 37 17

Also, I left a few comments (review process) on the errors raised for permissions and scopes as I was not sure about their messages.

PS.: the edit was to make the example more complete 😉

### CODE REVIEW

File: auth0/v3/authentication/token_verifier.py [hunk 1]
```diff
@@ -340,3 +339,130 @@ def _verify_payload(self, payload, nonce=None, max_age=None, organization=None):
...
+
+        # Permissions
+        if permissions:
+            if 'permissions' not in payload or not isinstance(payload['permissions'], list):
+                raise TokenValidationError('No permissions claim found in token.')
+
+            for permission in permissions:
+                if permission not in payload['permissions']:
+                    msg = "You don't have access to this resource"
+                    raise Auth0Error(status_code=403, error_code=403, message=msg)
```
Comment (jtemporal) [inline auth0/v3/authentication/token_verifier.py:468]: I'm not sure this is the correct error status code or error code and message for this case

File: auth0/v3/authentication/token_verifier.py [hunk 2]
```diff
@@ -340,3 +339,130 @@ def _verify_payload(self, payload, nonce=None, max_age=None, organization=None):
...
+
+        # Scopes
+        if scopes:
+            if 'scopes' not in payload or not isinstance(payload['scopes'], str):
+                raise TokenValidationError('No scopes claim found in token.')
+
+            for scope in scopes.split(' '):
+                if scope not in payload['scopes'].split(' '):
+                    msg = "You don't have access to this resource"
+                    raise Auth0Error(status_code=401, error_code=401, message=msg)
```
Comment (jtemporal) [inline auth0/v3/authentication/token_verifier.py:458]: I'm not sure this is the correct error status code or error code and message for this case

File: auth0/v3/authentication/token_verifier.py (hunk 1 above)
Comment (Sambego) [reply auth0/v3/authentication/token_verifier.py:468]: `403` Forbidden is correct when the user does not have the correct permissions

File: auth0/v3/authentication/token_verifier.py (hunk 2 above)
Comment (Sambego) [reply auth0/v3/authentication/token_verifier.py:458]: The OAuth spec specifies a `400` reqiest for an invalid scope error, something like: 

```
HTTP/1.1 400 Bad Request
     Content-Type: application/json;charset=UTF-8
     Cache-Control: no-store
     Pragma: no-cache

     {
       "error":"invalid_scope"
     }
```

https://datatracker.ietf.org/doc/html/rfc6749#section-5.2

Comment (Sambego) [reply auth0/v3/authentication/token_verifier.py]: While digging deeper in the OAuth 2.0 spec, it seems the above `400` needs to be thrown on requesting an access token with in invalid scopes, not when validating an access token's scopes.

The spec does not specify how to handle this scenario

> If a resource access request fails, the resource server SHOULD inform
   the client of the error.  While the specifics of such error responses
   are beyond the scope of this specification, this document establishes
   a common registry in Section 11.4 for error values to be shared among
   OAuth token authentication schemes. 

The `express-oauth2-bearer` lib [throws](https://github.com/auth0/express-oauth2-bearer/blob/master/lib/errors.js#L22) a `403`, let's use that to stay consistent.

File: auth0/v3/authentication/token_verifier.py [hunk 3]
```diff
@@ -340,3 +339,135 @@ def _verify_payload(self, payload, nonce=None, max_age=None, organization=None):
...
+
+    def verify(self, token, scopes=None, permissions=None):
+        """Attempts to verify the given JWT access token, following the steps
+        defined in the RFC 7519, for more details see:
+        https://datatracker.ietf.org/doc/html/rfc7519#section-7.2
+        Auth0 only uses JWT access tokens.
+
+        Args:
+            token (str): The JWT access token to verify.
+            scopes (str, optional): A string of scopes to verify.
```
Comment (lbalmaceda) [inline auth0/v3/authentication/token_verifier.py:378]: I find it a bit strange that this one is named "scopes" (plural) and the type is a string, versus "permissions" (also plural) where the type is an array of strings. I'd try to be consistent about it. And also, if we keep the "string" types like in scope, clarify how we expect them to be separated (using whitespace)

File: auth0/v3/authentication/token_verifier.py [hunk 4]
```diff
@@ -340,3 +339,135 @@ def _verify_payload(self, payload, nonce=None, max_age=None, organization=None):
...
+        # Issuer
+        if 'iss' not in payload or not isinstance(payload['iss'], (str, ustr)):
+            raise TokenValidationError('Issuer (iss) claim must be a string present in the access token')
+        if payload['iss'] != self.iss:
+            raise TokenValidationError(
+                'Issuer (iss) claim mismatch in the access token; expected "{}", '
+                'found "{}"'.format(self.iss, payload['iss']))
+
+        # Subject
+        if 'sub' not in payload or not isinstance(payload['sub'], (str, ustr)):
```
Comment (lbalmaceda) [inline auth0/v3/authentication/token_verifier.py:419]: where is this requirement coming from? I couldn't find any article/doc mentioning the `sub` claim should be validated when checking an access token. Only found signature, time-related claims, audience and issuer.

File: auth0/v3/authentication/token_verifier.py [hunk 5]
```diff
@@ -340,3 +339,135 @@ def _verify_payload(self, payload, nonce=None, max_age=None, organization=None):
...
+                'after expiration time ({})'.format(now, exp_time))
+
+        # Issued at
+        if 'iat' not in payload or not isinstance(payload['iat'], int):
+            raise TokenValidationError('Issued At (iat) claim must be a number present in the access token')
+
+        # Scopes
+        if scopes:
+            if 'scope' not in payload or not isinstance(payload['scope'], (str, ustr)):
+                raise TokenValidationError('No scopes claim found in token.')
```
Comment (lbalmaceda) [inline auth0/v3/authentication/token_verifier.py:458]: Is this claim required per the spec or Auth0's implementation? Seems like instead, we should only be failing if this is not set if we were given an expected value for it in the first place. 

The message should be similar to the rest. Something like `{claim} ({claim short name}) claim must be a {claim type} present in the access token`.

File: auth0/v3/authentication/token_verifier.py [hunk 6]
```diff
@@ -340,3 +339,135 @@ def _verify_payload(self, payload, nonce=None, max_age=None, organization=None):
...
+
+            for scope in scopes.split(' '):
+                if scope not in payload['scope'].split(' '):
+                    msg = "Insufficient scope."
+                    raise Auth0Error(status_code=403, error_code='insufficient_scope', message=msg)
+
+        # Permissions
+        if permissions:
+            if 'permissions' not in payload or not isinstance(payload['permissions'], list):
+                raise TokenValidationError('No permissions claim found in token.')
```
Comment (lbalmaceda) [inline auth0/v3/authentication/token_verifier.py:468]: Is this claim required per the spec or Auth0's implementation? Seems like instead, we should only be failing if this is not set if we were given an expected value for it in the first place. 

The message should be similar to the rest. Something like `{claim} ({claim short name}) claim must be a {claim type} present in the access token`.

File: auth0/v3/test/authentication/test_token_verifier.py [hunk 7]
```diff
@@ -402,4 +412,222 @@ def test_fails_when_org_specified_but_not_(self):
...
+            audience=expectations_2['audience'],
+        )
+        tv._clock = MOCKED_CLOCK
+        tv.verify(token, scopes='scope:two scope:one')
+
+    def test_fails_when_scopes_specified_but_not_present(self):
+        token = "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJhdXRoMHxzZGs0NThma3MiLCJhdWQiOiJ0b2tlbnMtdGVzdC0xMjMiLCJpc3MiOiJodHRwczovL3Rva2Vucy10ZXN0LmF1dGgwLmNvbS8iLCJleHAiOjE1ODc3NjUzNjEsImlhdCI6MTU4NzU5MjU2MX0.wotJnUdD5IfdZMewF_-BnHc0pI56uwzwr5qaSXvSu9w"
+        self.assert_fails_with_token_error(token, "No scopes claim found in token.", signature_verifier=SymmetricSignatureVerifier(HMAC_SHARED_SECRET), scopes='scope:three')
+
+    def test_fails_when_scopes_specified_but_does_not_mactch(self):
```
Comment (lbalmaceda) [inline auth0/v3/test/authentication/test_token_verifier.py:609]:
```suggestion
    def test_fails_when_scopes_specified_but_does_not_match(self):
```

File: auth0/v3/test/authentication/test_token_verifier.py [hunk 8]
```diff
@@ -402,4 +412,222 @@ def test_fails_when_org_specified_but_not_(self):
...
+            audience=expectations_2['audience'],
+        )
+        tv._clock = MOCKED_CLOCK
+        tv.verify(token, permissions=['scope:two'])
+
+    def test_fails_when_permissions_specified_but_not_present(self):
+        token = "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJhdXRoMHxzZGs0NThma3MiLCJhdWQiOiJ0b2tlbnMtdGVzdC0xMjMiLCJpc3MiOiJodHRwczovL3Rva2Vucy10ZXN0LmF1dGgwLmNvbS8iLCJleHAiOjE1ODc3NjUzNjEsImlhdCI6MTU4NzU5MjU2MX0.wotJnUdD5IfdZMewF_-BnHc0pI56uwzwr5qaSXvSu9w"
+        self.assert_fails_with_token_error(token, "No permissions claim found in token.", signature_verifier=SymmetricSignatureVerifier(HMAC_SHARED_SECRET), permissions=['scope:three'])
+
+    def test_fails_when_permissions_specified_but_does_not_mactch(self):
```
Comment (lbalmaceda) [inline auth0/v3/test/authentication/test_token_verifier.py:579]:
```suggestion
    def test_fails_when_permissions_specified_but_does_not_match(self):
```

**Review by lbalmaceda — CHANGES_REQUESTED** [review]
I skipped reviewing the test for now

**Comment by adamjmcgrath** [comment]
Thanks for your great contribution @jtemporal!

For now we're going to stick with recommending 3rd party libraries (like PyJWT) specifically for verifing Access Token JWTs, eg

```py
import jwt
from jwt import PyJWKClient

token = "xxx.xxx.xxx"
url = "https://foobar.auth0.com/.well-known/jwks.json"

jwks_client = PyJWKClient(url)
signing_key = jwks_client.get_signing_key_from_jwt(token)

data = jwt.decode(
    token,
    signing_key.key,
    algorithms=["RS256"],
    audience="https://expenses-api",
    issuer="https://foobar.auth0.com/"
)

# check data.scopes / data.permissions
# Issue errors per rfc6750 if you're protecting an API
```

But this is constantly under review

**Pull request closed without merging**